ydb/core/kqp/provider/yql_kikimr_datasource.cpp

```diff
@@ -673,58 +674,73 @@ class TKikimrDataSource : public TDataProviderBase {
+                                                .Add(ctx.NewCallable(node->Pos(), "MrTableConcat", {newKey}))
+                                                .Add(ctx.NewCallable(node->Pos(), "Void", {}))
+                                                .Add(BuildExternalTableSettings(node->Pos(), ctx, tableDesc.Metadata->Columns, source, tableDesc.Metadata->ExternalSource.TableContent))
+
+                                            .Build()
+                                            .Done().Ptr();
+                    auto retChildren = node->ChildrenList();
+                    retChildren[0] = newRead;
+                    return ctx.ChangeChildren(*node, std::move(retChildren));
+                }
+            } else if (tableDesc.Metadata->Kind == EKikimrTableKind::View) {
+                ctx.Step
+                    .Repeat(TExprStep::ExprEval)
+                    .Repeat(TExprStep::DiscoveryIO)
                     .Repeat(TExprStep::Epochs)
                     .Repeat(TExprStep::Intents)
                     .Repeat(TExprStep::LoadTablesMetadata)
                     .Repeat(TExprStep::RewriteIO);
-            TExprNode::TPtr path = ctx.NewCallable(node->Pos(), "String", { ctx.NewAtom(node->Pos(), tableDesc.Metadata->ExternalSource.TableLocation) });
-            auto table = ctx.NewList(node->Pos(), {ctx.NewAtom(node->Pos(), "table"), path});
-            auto newKey = ctx.NewCallable(node->Pos(), "Key", {table});
-            auto newRead = Build<TCoRead>(ctx, node->Pos())
-                                    .World(read->Child(0))
-                                    .DataSource(
-                                        Build<TCoDataSource>(ctx, node->Pos())
-                                            .Category(ctx.NewAtom(node->Pos(), source->GetName()))
-                                            .FreeArgs()
-                                                .Add(ctx.NewAtom(node->Pos(), tableDesc.Metadata->ExternalSource.DataSourcePath))
-                                            .Build()
-                                        .Done().Ptr()
-                                    )
-                                    .FreeArgs()
-                                        .Add(ctx.NewCallable(node->Pos(), "MrTableConcat", {newKey}))
-                                        .Add(ctx.NewCallable(node->Pos(), "Void", {}))
-                                        .Add(BuildExternalTableSettings(node->Pos(), ctx, tableDesc.Metadata->Columns, source, tableDesc.Metadata->ExternalSource.TableContent))
 
-                                    .Build()
-                                    .Done().Ptr();
-            auto retChildren = node->ChildrenList();
-            retChildren[0] = newRead;
-            return ctx.ChangeChildren(*node, std::move(retChildren));
+                const auto& query = tableDesc.Metadata->ViewPersistedData.QueryText;
+                return RewriteReadFromView(node, ctx, query, cluster);
```

Comment:
Forgot to add a feature flag check. I suppose it should go somewhere here. Will add in the next revision of this review.

Reply:
Added the feature flag check in the next revision